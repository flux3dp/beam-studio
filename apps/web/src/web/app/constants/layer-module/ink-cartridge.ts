export enum InkDetectionStatus {
  PENDING = 0, // The door was opened.
  FAILED = 1,
  SUCCESS = 2, // valid and used
  VALIDATE_FAILED = 3, // Secure chip validation failed.
  UNUSED = 4, // valid but not used
}

export default {};
