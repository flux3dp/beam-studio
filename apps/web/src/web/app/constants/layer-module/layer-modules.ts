// ref: https://www.notion.so/flux3dp/030619b6721849498cadc57e660107d3?pvs=4

enum LayerModule {
  LASER_10W_DIODE = 1, // or default CO2 laser for BeamSeries
  LASER_20W_DIODE = 2,
  LASER_1064 = 4,
  PRINTER = 5,
  UNKNOWN = 9,
  LASER_UNIVERSAL = 15,
}

export const modelsWithModules = new Set(['fad1', 'ado1']);

export default LayerModule;
