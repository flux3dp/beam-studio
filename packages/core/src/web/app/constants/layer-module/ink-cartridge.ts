export enum InkDetectionStatus {
  FAILED = 1,
  PENDING = 0, // The door was opened.
  SUCCESS = 2, // valid and used
  UNUSED = 4, // valid but not used
  VALIDATE_FAILED = 3, // Secure chip validation failed.
}

export default {};
