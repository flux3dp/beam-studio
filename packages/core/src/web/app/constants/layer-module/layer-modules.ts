// ref: https://www.notion.so/flux3dp/030619b6721849498cadc57e660107d3?pvs=4

/* eslint-disable perfectionist/sort-objects */
/**
 * Layer Module ID when exporting task
 */
export const LayerModule = {
  UV_PRINT: -1,
  LASER_10W_DIODE: 1, // or default CO2 laser for BeamSeries
  LASER_20W_DIODE: 2,
  LASER_1064: 4,
  PRINTER: 5,
  PRINTER_4C: 7,
  // virtual values for uv layer (they are seen as printer 4c in firmware)
  UV_WHITE_INK: 8,
  UV_VARNISH: 9,
  LASER_UNIVERSAL: 15,
} as const;

/**
 * Layer Module ID reporting from device info (used in device info display)
 */
export const DetectedLayerModule = {
  NONE: 0, // placeholder for device info search
  LASER_10W_DIODE: 1, // or default CO2 laser for BeamSeries
  LASER_20W_DIODE: 2,
  LASER_1064: 4,
  PRINTER: 5,
  PRINTER_4C: 7,
  PRINTER_4C_WITH_1064: 8,
  UNKNOWN: 9,
  PRINTER_4C_WITH_UV: 10,
  PRINTER_4C_WITH_UV_1064: 11,
} as const;
/* eslint-enable perfectionist/sort-objects */
export type LayerModuleType = (typeof LayerModule)[keyof typeof LayerModule];
export type DetectedLayerModuleType = (typeof DetectedLayerModule)[keyof typeof DetectedLayerModule];

export const printingModules = new Set<LayerModuleType>([LayerModule.PRINTER, LayerModule.PRINTER_4C]);

export const fullColorModulesArray = [LayerModule.PRINTER, LayerModule.PRINTER_4C, LayerModule.UV_PRINT] as const;
export const fullColorModules = new Set<LayerModuleType>(fullColorModulesArray);
// uv for fbm2, not for uv printing export
export const UVModules = new Set<LayerModuleType>([LayerModule.UV_WHITE_INK, LayerModule.UV_VARNISH]);
