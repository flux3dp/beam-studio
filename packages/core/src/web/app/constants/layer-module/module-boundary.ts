import LayerModule from 'app/constants/layer-module/layer-modules';

// Module Boundary for ador
// bottom boundary is added by 20mm for 1064nm laser
const moduleBoundary: {
  [model: number]: { top: number; left: number; bottom: number; right: number };
} = {
  [LayerModule.LASER_10W_DIODE]: { top: 0, left: 0, bottom: 20, right: 0 },
  [LayerModule.LASER_20W_DIODE]: { top: 0, left: 0, bottom: 30, right: 0 },
  [LayerModule.LASER_1064]: { top: 0, left: 0, bottom: 38, right: 0 },
  [LayerModule.PRINTER]: { top: 12.7, left: 0, bottom: 50, right: 0 },
};

export default moduleBoundary;
