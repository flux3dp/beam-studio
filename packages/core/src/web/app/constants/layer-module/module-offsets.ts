// Default value of module offset
import LayerModule from 'app/constants/layer-module/layer-modules';

// Module Boundary for ador
const moduleOffsets = {
  [LayerModule.LASER_10W_DIODE]: [0, 0],
  [LayerModule.LASER_20W_DIODE]: [0, 0],
  [LayerModule.LASER_1064]: [0, 26.95],
  [LayerModule.PRINTER]: [0, -13.37],
  [LayerModule.LASER_UNIVERSAL]: [0, 0],
};

export default moduleOffsets;
