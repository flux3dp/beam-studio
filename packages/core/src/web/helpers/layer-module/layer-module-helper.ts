import beamboxPreference from 'app/actions/beambox/beambox-preference';
import LayerModule from 'app/constants/layer-module/layer-modules';
import i18n from 'helpers/i18n';

const LaserModuleSet = new Set([
  LayerModule.LASER_10W_DIODE,
  LayerModule.LASER_20W_DIODE,
]);

const getDefaultLaserModule = (): LayerModule => {
  const value = beamboxPreference.read('default-laser-module') as LayerModule;
  if (LaserModuleSet.has(value)) return value;
  return LayerModule.LASER_20W_DIODE;
}

const getModulesTranslations = (): { [module: number]: string } => {
  const t = i18n.lang.layer_module;
  return {
    0: t.none,
    [LayerModule.LASER_10W_DIODE]: t.laser_10w_diode,
    [LayerModule.LASER_20W_DIODE]: t.laser_20w_diode,
    [LayerModule.LASER_1064]: t.laser_2w_infrared,
    [LayerModule.PRINTER]: t.printing,
    [LayerModule.UNKNOWN]: t.unknown,
  };
};

export default {
  getDefaultLaserModule,
  getModulesTranslations,
};
