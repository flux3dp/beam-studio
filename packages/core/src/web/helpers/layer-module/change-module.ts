import alertCaller from '@core/app/actions/alert-caller';
import presprayArea from '@core/app/actions/canvas/prespray-area';
import alertConstants from '@core/app/constants/alert-constants';
import type { LayerModuleType } from '@core/app/constants/layer-module/layer-modules';
import { LayerModule, printingModules } from '@core/app/constants/layer-module/layer-modules';
import { useDocumentStore } from '@core/app/stores/documentStore';
import history from '@core/app/svgedit/history/history';
import undoManager from '@core/app/svgedit/history/undoManager';
import initLayerConfigState from '@core/app/views/beambox/Right-Panels/ConfigPanel/initState';
import LayerPanelController from '@core/app/views/beambox/Right-Panels/contexts/LayerPanelController';
import alertConfig from '@core/helpers/api/alert-config';
import i18n from '@core/helpers/i18n';
import toggleFullColorLayer from '@core/helpers/layer/full-color/toggleFullColorLayer';
import {
  applyPreset,
  baseConfig,
  getData,
  moduleBaseConfig,
  writeDataLayer,
} from '@core/helpers/layer/layer-config-helper';
import presetHelper from '@core/helpers/presets/preset-helper';

export const changeLayersModule = async (
  layers: Element[],
  oldValue: LayerModuleType,
  newValue: LayerModuleType,
  { addToHistory = false }: { addToHistory?: boolean } = {},
): Promise<boolean> => {
  const workarea = useDocumentStore.getState().workarea;
  const isCurrentPrinting = printingModules.has(oldValue);
  const isChangingToPrinting = printingModules.has(newValue);

  if (isCurrentPrinting && !isChangingToPrinting && !alertConfig.read('skip-switch-to-printer-module')) {
    const res = await new Promise((resolve) => {
      alertCaller.popUp({
        buttonType: alertConstants.CONFIRM_CANCEL,
        caption: i18n.lang.layer_module.notification.convertFromPrintingModuleTitle,
        checkbox: {
          callbacks: [
            () => {
              alertConfig.write('skip-switch-to-printer-module', true);
              resolve(true);
            },
            () => resolve(false),
          ],
          text: i18n.lang.alert.dont_show_again,
        },
        id: 'switch-to-printer-module',
        message: i18n.lang.layer_module.notification.convertFromPrintingModuleMsg,
        messageIcon: 'notice',
        onCancel: () => resolve(false),
        onConfirm: () => resolve(true),
      });
    });

    if (!res) {
      return false;
    }
  } else if (!isCurrentPrinting && isChangingToPrinting && !alertConfig.read('skip-switch-to-laser-module')) {
    const res = await new Promise((resolve) => {
      alertCaller.popUp({
        buttonType: alertConstants.CONFIRM_CANCEL,
        caption: i18n.lang.layer_module.notification.convertFromLaserModuleTitle,
        checkbox: {
          callbacks: [
            () => {
              alertConfig.write('skip-switch-to-laser-module', true);
              resolve(true);
            },
            () => resolve(false),
          ],
          text: i18n.lang.alert.dont_show_again,
        },
        id: 'switch-to-laser-module',
        message: i18n.lang.layer_module.notification.convertFromLaserModuleMsg,
        messageIcon: 'notice',
        onCancel: () => resolve(false),
        onConfirm: () => resolve(true),
      });
    });

    if (!res) {
      return false;
    }
  }

  const presetsList = presetHelper.getPresetsList(workarea, oldValue);
  const newPresetsList = presetHelper.getPresetsList(workarea, newValue);
  const batchCmd = new history.BatchCommand('Change layer module');
  const isBM2IR = workarea === 'fbm2' && newValue === LayerModule.LASER_1064;

  layers.forEach((layer) => {
    const configName = getData(layer, 'configName');
    const repeat = getData(layer, 'repeat');
    const oldPreset = configName ? presetsList.find(({ key, name }) => [key, name].includes(configName)) : null;
    const newPreset = oldPreset ? newPresetsList.find(({ key, name }) => [key, name].includes(configName)) : null;

    writeDataLayer(layer, 'module', newValue, { batchCmd });

    if (newValue === LayerModule.UV_PRINT) writeDataLayer(layer, 'repeat', 0, { batchCmd });
    else if (repeat === 0) writeDataLayer(layer, 'repeat', 1, { batchCmd });

    if (!newPreset) {
      writeDataLayer(layer, 'configName', undefined, { batchCmd });

      if (isCurrentPrinting && !isChangingToPrinting) {
        writeDataLayer(layer, 'speed', baseConfig.speed, { batchCmd });
        writeDataLayer(layer, 'power', baseConfig.power, { batchCmd });
      } else if (!isCurrentPrinting && isChangingToPrinting) {
        writeDataLayer(layer, 'printingSpeed', moduleBaseConfig[newValue]?.printingSpeed ?? baseConfig.printingSpeed, {
          batchCmd,
        });
        writeDataLayer(layer, 'ink', moduleBaseConfig[newValue]?.ink ?? baseConfig.ink, { batchCmd });
        writeDataLayer(layer, 'multipass', moduleBaseConfig[newValue]?.multipass ?? baseConfig.multipass, { batchCmd });
      } else if (isBM2IR) {
        const currentSpeed = getData(layer, 'speed');

        if (currentSpeed && currentSpeed > 150) writeDataLayer(layer, 'speed', 150, { batchCmd });
      }
    } else if (newPreset !== oldPreset) {
      applyPreset(layer, newPreset, { batchCmd });
    }

    const toggleFullColorCmd = toggleFullColorLayer(layer, { val: isChangingToPrinting });

    if (toggleFullColorCmd && !toggleFullColorCmd.isEmpty()) batchCmd.addSubCommand(toggleFullColorCmd);
  });

  initLayerConfigState();
  LayerPanelController.updateLayerPanel();
  presprayArea.togglePresprayArea();

  if (addToHistory) {
    batchCmd.onAfter = () => {
      initLayerConfigState();
      LayerPanelController.updateLayerPanel();
      presprayArea.togglePresprayArea();
    };
    undoManager.addCommandToHistory(batchCmd);
  }

  return true;
};
