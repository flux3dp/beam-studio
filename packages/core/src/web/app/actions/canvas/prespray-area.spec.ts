const mockGetState = jest.fn();

jest.mock('@core/app/stores/documentStore', () => ({
  useDocumentStore: {
    getState: () => mockGetState(),
    subscribe: () => jest.fn(),
  },
}));

import { LayerModule, type LayerModuleType } from '@core/app/constants/layer-module/layer-modules';
import presprayArea from './prespray-area';

const mockGetWidth = jest.fn();
const mockGetHeight = jest.fn();
const mockGetExpansion = jest.fn();
const mockGetMinY = jest.fn();
const mockGetModel = jest.fn();

jest.mock('@core/app/svgedit/workarea', () => ({
  get expansion() {
    return mockGetExpansion();
  },
  get height() {
    return mockGetHeight();
  },
  get minY() {
    return mockGetMinY();
  },
  get model() {
    return mockGetModel();
  },
  get modelHeight() {
    return mockGetHeight();
  },
  get width() {
    return mockGetWidth();
  },
}));

const mockAddCommandToHistory = jest.fn();

jest.mock('@core/app/svgedit/history/undoManager', () => ({
  addCommandToHistory: (...args) => mockAddCommandToHistory(...args),
}));

const mockHasModuleLayer = jest.fn();

jest.mock('@core/helpers/layer-module/layer-module-helper', () => ({
  hasModuleLayer: (...args) => mockHasModuleLayer(...args),
}));

const setHasModuleLayer = (value: LayerModuleType | null) => {
  if (value === null) {
    mockHasModuleLayer.mockReturnValue(false);
  } else {
    mockHasModuleLayer.mockImplementation((modules: LayerModuleType[]) => modules.includes(value));
  }
};
const mockRequestAnimationFrame = jest.fn();

describe('test canvas/prespray-area', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    mockGetWidth.mockReturnValue(4300);
    mockGetHeight.mockReturnValue(4000);
    mockGetMinY.mockReturnValue(0);
    mockGetModel.mockReturnValue('ado1');
    mockGetExpansion.mockReturnValue([0, 0]);
    mockRequestAnimationFrame.mockImplementation((cb) => cb());
    mockGetState.mockReturnValue({
      'enable-4c-prespray-area': false,
      'enable-job-origin': false,
      rotary_mode: false,
    });
    mockHasModuleLayer.mockReturnValue(false);
    window.requestAnimationFrame = mockRequestAnimationFrame;
    document.body.innerHTML = '<svg id="fixedSizeSvg"></svg>';
    setHasModuleLayer(LayerModule.PRINTER);
  });

  test('generate prespray area for PRINTER', () => {
    presprayArea.generatePresprayArea();
    expect(document.getElementById('fixedSizeSvg').innerHTML).toMatchSnapshot();
    expect(document.getElementById('presprayAreaImage')).not.toBeNull();
    expect(document.getElementById('presprayAreaRect')).not.toBeNull();
    expect(document.getElementById('presprayAreaImage').getAttribute('display')).not.toBe('none');
    expect(document.getElementById('presprayAreaRect').getAttribute('display')).toBe('none');
  });

  test('generate prespray area for PRINTER_4C', () => {
    mockGetModel.mockReturnValue('fbm2');
    setHasModuleLayer(LayerModule.PRINTER_4C);
    mockGetState.mockReturnValue({
      'enable-4c-prespray-area': true,
      'enable-job-origin': false,
      rotary_mode: false,
    });
    presprayArea.generatePresprayArea();
    expect(document.getElementById('presprayAreaImage')).not.toBeNull();
    expect(document.getElementById('presprayAreaRect')).not.toBeNull();
    expect(document.getElementById('presprayAreaImage').getAttribute('display')).toBe('none');
    expect(document.getElementById('presprayAreaRect').getAttribute('display')).not.toBe('none');
  });

  test('hide prespray area when rotary mode is on', () => {
    mockGetState.mockReturnValue({
      'enable-job-origin': false,
      rotary_mode: true,
    });
    presprayArea.generatePresprayArea();
    expect(document.getElementById('presprayAreaImage')).not.toBeNull();
    expect(document.getElementById('presprayAreaRect')).not.toBeNull();
    expect(document.getElementById('presprayAreaImage').getAttribute('display')).toBe('none');
    expect(document.getElementById('presprayAreaRect').getAttribute('display')).toBe('none');
  });

  test('toggle prespray area', () => {
    presprayArea.generatePresprayArea();
    expect(document.getElementById('presprayAreaImage')).not.toBeNull();
    expect(document.getElementById('presprayAreaRect')).not.toBeNull();
    setHasModuleLayer(null);
    presprayArea.togglePresprayArea();
    expect(document.getElementById('presprayAreaImage').getAttribute('display')).toBe('none');
    expect(document.getElementById('presprayAreaRect').getAttribute('display')).toBe('none');
  });

  test('checkMouseTarget', () => {
    presprayArea.generatePresprayArea();

    const imageTarget = document.getElementById('presprayAreaImage');
    const rectTarget = document.getElementById('presprayAreaRect');

    expect(presprayArea.checkMouseTarget(imageTarget)).toBe(true);
    expect(presprayArea.checkMouseTarget(rectTarget)).toBe(true);

    const otherTarget = document.createElementNS('http://www.w3.org/2000/svg', 'rect');

    expect(presprayArea.checkMouseTarget(otherTarget)).toBe(false);
  });

  test('drag prespray area for PRINTER (free movement)', () => {
    presprayArea.generatePresprayArea();
    expect(presprayArea.getPosition()).toEqual({ h: 300, w: 300, x: 4000, y: 2400 });
    presprayArea.startDrag();
    presprayArea.drag(-1000, -1000);
    expect(presprayArea.getPosition()).toEqual({ h: 300, w: 300, x: 3000, y: 1400 });
    presprayArea.drag(-3000, -3000);
    expect(presprayArea.getPosition()).toEqual({ h: 300, w: 300, x: 1000, y: 0 });
    presprayArea.startDrag();
    presprayArea.drag(100, 100);
    expect(presprayArea.getPosition()).toEqual({ h: 300, w: 300, x: 1100, y: 100 });
    presprayArea.drag(10000, 10000);
    expect(presprayArea.getPosition()).toEqual({ h: 300, w: 300, x: 4000, y: 3700 });
  });

  test('drag prespray area for PRINTER_4C (horizontal only)', () => {
    setHasModuleLayer(LayerModule.PRINTER_4C);
    mockGetModel.mockReturnValue('fbm2');
    mockGetState.mockReturnValue({
      'enable-4c-prespray-area': true,
      'enable-job-origin': false,
      rotary_mode: false,
    });
    document.body.innerHTML = '<svg id="fixedSizeSvg"></svg>';
    presprayArea.generatePresprayArea();
    expect(presprayArea.getPosition()).toEqual({ h: 0, w: 223, x: 4077, y: 0 });
    presprayArea.startDrag();
    presprayArea.drag(-1000, -1000);
    // Y should remain at 0 for PRINTER_4C
    expect(presprayArea.getPosition()).toEqual({ h: 0, w: 223, x: 3077, y: 0 });
    presprayArea.drag(-3000, -3000);
    expect(presprayArea.getPosition()).toEqual({ h: 0, w: 223, x: 1077, y: 0 });
    presprayArea.startDrag();
    presprayArea.drag(100, 100);
    // Y should still remain at 0
    expect(presprayArea.getPosition()).toEqual({ h: 0, w: 223, x: 1177, y: 0 });
    presprayArea.drag(10000, 10000);
    expect(presprayArea.getPosition()).toEqual({ h: 0, w: 223, x: 4077, y: 0 });
  });
});
