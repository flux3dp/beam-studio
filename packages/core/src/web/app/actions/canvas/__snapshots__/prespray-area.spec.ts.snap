// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test canvas/prespray-area generate prespray area 1`] = `"<g class="layer" data-module="5"></g><image id="presprayArea" x="4000" y="2400" width="300" height="300" href="url" style="cursor:move;"><title>Priming Area</title></image>"`;
